src/DataProtection/DataProtection/src/TypeExtensions.cs

```diff
@@ -39,4 +40,15 @@ public static Type GetTypeWithTrimFriendlyErrorMessage(string typeName)
             throw new InvalidOperationException($"Unable to load type '{typeName}'. If the app is published with trimming then this type may have been trimmed. Ensure the type's assembly is excluded from trimming.", ex);
         }
     }
+
+    public static bool MatchType(string resolvedTypeName, ITypeNameResolver typeNameResolver, Type matchType)
+    {
+        // Before attempting to resolve the name to a type, check if it starts with the full name of the type.
+        if (matchType.FullName != null && resolvedTypeName.StartsWith(matchType.FullName, StringComparison.Ordinal))
```

Comment:
> Oh, maybe `resolvedTypeName` is expected to be assembly-qualified? (And we don't want compare it to `matchType.AssemblyQualifiedName` because of perf or because it might differ or something?)

Yes. I'll add a comment.

> Do we care about the case where the type is a closed generic, in which case the type arguments will be assembly-qualified?

I don't think any of the types we're comparing against has closed generics. I'll double check.